Switching external resources to be device-local only.

benvanik opened the pull request:
Previously all external resources (results returned by an invocation) were made host-visible and mappable and this prevented the use of queue-ordered allocations in CUDA as memory pools cannot service memory with associated host pointers. Depending on device the host-visible memory could also be much slower to access (or have more potential pitfalls with page management) vs pinned device-local memory and this got worse once we started doing more dispatches in-place on the results.

Now all external buffers are by default allocated as device-local. Users will need to manually stage the buffers and otherwise they'll remain on-device. For externalized state this is a good thing as it means we'll keep state on device automatically. A temporary flag has been added to revert to the old mappable behavior with `--iree-stream-external-resources-mappable=true`. Note that some devices (like CPU) will always allow mapping even if not requested and users can avoid the copies by checking before performing the transfers.

GPT2 CUDA post-change with alloca and no caching allocator enabled (~5us/invocation allocation overhead):
![image](https://github.com/openxla/iree/assets/75337/5f7f589d-b602-49b3-96c6-5c9dfa6578fe)

ScottTodd commented on build_tools/cmake/iree_benchmark_suite.cmake:
```diff
@@ -410,7 +410,6 @@ function(iree_benchmark_suite)
       # the generated artifacts.
       set(_FLAG_FILE "${_RUN_SPEC_DIR}/flagfile")
       set(_ADDITIONAL_ARGS "${_RULE_RUNTIME_FLAGS}")
-      list(APPEND _ADDITIONAL_ARGS "--device_allocator=caching")
```
Reminder to revert https://github.com/openxla/iree/pull/14016/commits/c9119d18d06f1dcefa92c34abc3f3d64d7726ff2 before merge

ScottTodd commented on compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp:
```diff
@@ -601,22 +601,119 @@ static void printWorkgroupCountRegion(OpAsmPrinter &p, Operation *op,
+  // guaranteed: a very low maximum buffer range may lead to packed slabs
+  // that are not fully addressable. For now we are processing models with
+  // small enough workloads and our target devices are relatively lax on
+  // things so long as we stay under UINT32_MAX boundaries.
```
> For now we are processing models with small enough workloads

Are we? We're seeing functions with 100s of arguments and 100GB+ of constants, right?

Maybe add to the comment what the code should be doing if the assumed upper limits are exceeded.

benvanik committed:
Switching external resources to be device-local only.
Previously all external resources (results returned by an invocation)
were made host-visible and mappable and this prevented the use of
queue-ordered allocations in CUDA as memory pools cannot service memory
with associated host pointers. Depending on device the host-visible
memory could also be much slower to access (or have more potential
pitfalls with page management) vs pinned device-local memory and this
got worse once we started doing more dispatches in-place on the results.

Now all external buffers are by default allocated as device-local. Users
will need to manually stage the buffers and otherwise they'll remain
on-device. For externalized state this is a good thing as it means we'll
keep state on device automatically. A temporary flag has been added to
revert to the old mappable behavior with
`--iree-stream-external-resources-mappable=true`. Note that some devices
(like CPU) will always allow mapping even if not requested and users can
avoid the copies by checking before performing the transfers.

raikonenfnu reviewed (approved):
LGTM! :)

Disclaimer: I don't fully understand the stream analysis part. but the code looks reasonable on removing assumed bits when the lifetime type is contradicting those bits if they were on the assumedBits.
